### PR TITLE
fix: resolve TypeScript error in PDFDownloadLink component react-pdf v4

### LINKED ISSUE #2886: TypeScript Error in PDFDownloadLink with React-PDF v4
**Describe the bug**
After updating to react-pdf v4, I encountered a TypeScript error when using the **PDFDownloadLink** component. The error occurs when passing a render prop to PDFDownloadLink, specifically when destructuring the **BlobProviderParams**.

**The following TypeScript error is shown:**
```bash
TS2769: No overload matches this call.
Overload 1 of 2,
(props: PDFDownloadLinkProps): PDFDownloadLink,
gave the following error.
Type ({ loading }: BlobProviderParams) => JSX.Element
is not assignable to type ReactNode | ReactElement<BlobProviderParams, string | JSXElementConstructor<any>>.
```
**To Reproduce**
1. Install or update to react-pdf v4.
2. Use the PDFDownloadLink component with a render prop that returns JSX.
3. Destructure the loading prop from the render function.
4. TypeScript throws the above error.

```jsx
<PDFDownloadLink
  className="text-decoration-none"
  document={<InvoicePdf invoice={invoice} hideDescription={hideDescription} />}
  fileName={`Invoice-${invoice?.invoiceID}.pdf`}
>
  {({ loading }) => (
    <SpinnerButton spin={loading} variant="outline-dark" size="sm">
      <div>
        Download Invoice
        <i className="bi bi-download ms-1" />
      </div>
    </SpinnerButton>
  )}
</PDFDownloadLink>;
```
**Expected behavior**
The PDFDownloadLink component should work as expected and correctly handle the render prop with the **BlobProviderParams** state, without any TypeScript errors.

**Desktop (please complete the following information):**

- OS: Windows
- Browser edge
- React-pdf version 4.0.0

### PR DESCRIPTION
**Pull request by islam-kamel**
Fixed #2886

**Changes Made:**
- Updated PDFDownloadLinkProps to correctly handle the 'children' prop, allowing for both ReactNode and function types.
- Ensured compatibility with TypeScript type definitions to prevent type mismatches.
- Added necessary type assertions to preserve functionality.

### CODE REVIEW

**Comment by robbycarter** [comment]
Anything i can do to help speed up this PR?

**Comment by mamlzy** [comment]
@diegomura sorry for tagging you

**Comment by fatihgnc** [comment]
I need this PR to be merged...

File: packages/renderer/index.d.ts [hunk 1]
```diff
@@ -495,12 +495,15 @@ declare namespace ReactPDF {
    */
   export class PDFViewer extends React.Component<PDFViewerProps> {}
 
+  // @ts-ignore
```
Comment (diegomura) [inline packages/renderer/index.d.ts:498]: Why is this needed?

**Comment by fabww** [comment]
This is still an issue in V4.1.5, is there any fix or workaround for this?

**Comment by mamlzy** [comment]
> This is still an issue in V4.1.5, is there any fix or workaround for this?

yes i still facing this issue. still using `@ts-expect-error` to ignoring the type errors

**Comment by diegomura** [comment]
There's an open question in this PR I left on code

**Review by goldenaca — APPROVED** [review]
OKAY

**Comment by Maquinours** [comment]
Up.
It is a big issue for all typescript users.

**Comment by mouchegmouradian** [comment]
Until it's fixed you can do something like this:

Screenshot 2024-12-17 at 11 45 46

the `@ts-expect-error` is better than `@ts-ignore` because when this PR is merged and the error isn't there anymore the `@ts-expect-error` will throw an error itself saying there are no errors, then you can remove it from your code safely.

**Comment by Maquinours** [comment]
Thank you !
